File: src/main/java/org/zibble/discordmessenger/commands/CommandFramework.java
package org.zibble.discordmessenger.commands;

import org.zibble.discordmessenger.components.readable.ReceivedCommand;

import java.util.Arrays;
import java.util.List;
import java.util.Optional;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.atomic.AtomicInteger;

public class CommandFramework {

    private static CommandFramework instance;
    private final List<DiscordCommand> commands = new CopyOnWriteArrayList<>();
    private final ExecutorService executor = Executors.newFixedThreadPool(6);

    public CommandFramework() {
        instance = this;
    }

    public void registerCommand(DiscordCommand command) {
        this.commands.add(command);
    }

    public void unregisterCommand(DiscordCommand command) {
        this.commands.remove(command);
    }

    public void runCommand(ReceivedCommand command) {
        executor.submit(() -> {
            for (DiscordCommand c : commands) {
                if (command.getContent().startsWith(c.getCommand())
                        || Arrays.stream(c.getAliases()).anyMatch(s -> command.getContent().startsWith(s))) {
                    try {
                        Command cmd = c;
                        String[] args = command.getContent().trim().split(" ");
                        if (args.length > 1 && cmd.getSubCommands() != null && c.getSubCommands().length > 0) {
                            AtomicInteger i = new AtomicInteger(1);
                            while (true) {
                                Optional<SubCommand> first = Arrays.stream(cmd.getSubCommands()).filter(s -> s.name().equalsIgnoreCase(args[i.get()])).findFirst();
                                if (first.isPresent()) {
                                    cmd = first.get();
                                    if (i.incrementAndGet() >= args.length)
                                        break;
                                } else {
                                    break;
                                }
                            }
                        }
                        cmd.execute(command);
                    } catch (Exception e) {
                        e.printStackTrace();
                    }
                    break;
                }
            }
        });
    }

    public List<DiscordCommand> getCommands() {
        return commands;
    }

    public static CommandFramework getInstance() {
        return instance;
    }

}

File: src/main/kotlin/org/zibble/discordmessenger/interaction/commands/CommandFramework.kt
package org.zibble.discordmessenger.interaction.commands

import org.zibble.discordmessenger.components.messagable.LegacyCommand
import org.zibble.discordmessenger.components.messagable.SlashCommand
import java.util.concurrent.CopyOnWriteArrayList
import java.util.concurrent.atomic.AtomicInteger

object CommandFramework {

    private val legacyCommands: MutableList<DiscordCommand<LegacyCommand>> = CopyOnWriteArrayList()
    private val commands: MutableList<DiscordCommand<SlashCommand>> = CopyOnWriteArrayList()

    fun registerLegacyCommand(command: DiscordCommand<LegacyCommand>) {
        legacyCommands.add(command)
    }

    fun unregisterLegacyCommand(command: DiscordCommand<LegacyCommand>) {
        legacyCommands.remove(command)
    }

    fun registerSlashCommand(command: DiscordCommand<SlashCommand>) {
        commands.add(command)
    }

    fun unregisterSlashCommand(command: DiscordCommand<SlashCommand>) {
        commands.remove(command)
    }

    suspend fun runCommand(command: org.zibble.discordmessenger.components.messagable.Command) {
        if (command is LegacyCommand) {
            legacyCommands.filter {
                command.content.startsWith(it.getCommand(), true) || it.getAliases()
                    .any { alias -> command.content.startsWith(alias, true) }
            }.forEach {
                var cmd: Command<LegacyCommand> = it
                val args: List<String> = command.content.trim().split(" ")
                if (args.size > 1 && it.getSubCommands().isNotEmpty()) {
                    val i = AtomicInteger(1)
                    while (true) {
                        val first = cmd.getSubCommands().firstOrNull { s -> s.name().equals(args[i.get()], true) }
                        if (first != null) {
                            cmd = first
                            if (i.incrementAndGet() >= args.size) break
                        } else {
                            break
                        }
                    }
                }
                cmd.execute(command)
            }
        } else {
            command as SlashCommand
            commands.filter {
                command.name.startsWith(it.getCommand(), true) || it.getAliases()
                    .any { alias -> command.name.startsWith(alias, true) }
            }.forEach {
                it.execute(command)
            }
        }
    }
}